lib/lure/repo.go

```diff
@@ -183,3 +190,132 @@ func Execute(pwd string, command string, params ...string) (string, error) {
 
 	return out, nil
 }
+
+func cleanupUpdateBranches(auth Authentication, project Project, repo Repo) error {
+
+	if project.Vcs != Hg {
+		return nil
+	}
+
+	trashBranch := project.TrashBranch
+	if trashBranch == "" {
+		log.Printf("Info: Project has no trash branch defined. Skipping cleanup.")
+		return nil
+	}
+
+	log.Printf("Info: Cleaning up lure branches with no associated PRs.")
+
+	out, err := repo.Cmd("branches", "--active", "--template", "{branches}\n")
+	if err != nil {
+		return err
+	}
+
+	branches := strings.Split(out, "\n")
+	branchPrefix := project.BranchPrefix
+	if branchPrefix == "" {
+		branchPrefix = "lure-"
+	}
+	msgRegex, err := regexp.Compile(`Update (\S*) to (\S*)`)
+	if err != nil {
+		return err
+	}
+
+	existingPRs := getPullRequests(auth, project.Owner, project.Name, true)
+
+	for _, branch := range branches {
+		if strings.HasPrefix(branch, branchPrefix) {
+
+			dead, err := isBranchDead(repo, branch, existingPRs, msgRegex)
+			if err != nil {
+				continue
+			}
+
+			if dead {
+				closeDeadBranch(repo, branch, trashBranch)
+			}
+		}
+	}
+
+	if os.Getenv("DRY_RUN") == "1" {
+		log.Println("Running in DryRun mode, not doing the pull request nor pushing the changes")
+	} else {
+		if _, err := repo.Push(); err != nil {
+			return err
+		}
+	}
+
+	log.Printf("Info: Lure branches clean up done.")
+
+	return nil
+}
+
+func isBranchDead(repo Repo, branch string, existingPRs []PullRequest, msgRegex *regexp.Regexp) (bool, error) {
```

Comment:
If I understand correctly, the goal is to close all branches that no longer have an associated PR. Have you considered fetching the list of branches (https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/refs/branches) and the list of PRs (https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/pullrequests) and remove branches that do not figure in the `source.branch` from those PRs? I think this solution would have the advantage of being more resilient then doing string matching.

Disclaimer: I haven't tryed it, so there might be some cavieat :/

Reply:
Not sure I get your solution correctly: won't this delete all branches people are working on and not in PR right now?

Reply:
Well we could filter for branches opened by lure?

Reply:
Yes, I did not think of using source.branch . Sounds like a much better solution.